File: DFS-BFS-3-changeWord/changeWord5.py
import random
import time
from collections import deque as queue

'''
두 단어에 대해 한 글자만 다르면 True 리턴하는 함수 간결하게 구현, (zip, 컴프리헨션)
큐로 구현 (단어, 단계 카운트)
words의 각 단어마다 다음 단계 노드들을  사전에 추가 {단어, (담 단계 노드들)}
단어 갯수보다 단계 카운트가 크면 못찾은거니까 0 리턴 ( ->, <- 반복으로 큐 안 빌 수 있음을 방지)
해당 노드와 연결된 다음 단계 노드들에 대해서 타겟과 같으면 단계 카운트 값을 리턴함
다음 단계 노드들을 큐에 추가함 
'''


def transistable(a, b): return sum((1 if x != y else 0)       # 두 단어를 비교하여 한 글자만 다르면 True 리턴
                                   for x, y in zip(a, b)) == 1


def solution(begin, target, words):
    q, d = queue(), dict()    # 큐와 사전 초기화
    q.append((begin, 0))      # 큐에 단어와 카운트 값 넣음
    # 첫 노드와 연결된 다음 단계 노드들, 해당 단어를 키값으로  다음 단계노드들을 set 밸류로 사전 추가
    d[begin] = set(filter(lambda x: transistable(x, begin), words))
    for w in words:                                                 # 각 단어마다
        # 다음 단계 노드들을 위와같이 사전에 추가
        d[w] = set(filter(lambda x: transistable(x, w), words))
    while q:
        cur, level = q.popleft()    # 큐에서 해당 노드와 단계 카운트값 불러옴
        if level > len(words):     # 단어 갯수보다 단계 카운트가 크면 못찾은거니까 리턴 0
            return 0
        for w in d[cur]:            # 해당 노드와 연결된 다음 단계 노드들에 대해서
            if w == target:         # 타겟과 같으면  단계 카운트 값을 리턴함
                return level + 1
            else:
                q.append((w, level + 1))   # 다음 단계 노드들을 큐에 추가함
    else:
        return 0


len_word = random.randint(3, 10)  # 각 단어의 길이는 3 이상 10 이하이며 모든 단어의 길이는 같음
words_num = random.randint(3, 50)  # words에는 3 ~ 50 개 단어가 있으며 중복 없음
words_num = 50
print("len_word: ", len_word)

# 각 단어는 길이가 모두 같고 알파벳 소문자로 구성됨 (26개 아스키 97 ~ 122)
# begin 만들기
begin = ""
for _ in range(len_word):
    begin += chr(random.randint(97, 122))
print("begin: ", begin)

# target 만들기 (단, begin과 같으면 안됨)
target = ""
while True:
    for _ in range(len_word):
        target += chr(random.randint(97, 122))
    if target != begin:
        break
print("target: ", target)

# words 만들기, 중복되는 단어 없음
words = []
cnt = 0
while True:
    word = ""
    for _ in range(len_word):
        word += chr(random.randint(97, 122))
    # print(len_word, word)
    if word in words:  # 단어가 중복되면 다시 단어 생성함
        continue
    words.append(word)
    cnt += 1
    if cnt == words_num:
        break
print(words_num, words)

# begin = "hit"
# target = "cog"
# words = ["hot", "dot", "dog", "lot", "log", "cog"]


# 실행 시간 및 결과 보기
start_time = time.time()
result = solution(begin, target, words)
print("--- %s seconds ---" % (time.time() - start_time))
print(result)

File: DFS-BFS-3-changeWord/changeWord5.cpp
#include <bits/stdc++.h>

using namespace std;

/*
dfs
visited

다음 노드를 찾이 못하면 dfs 진입 못함
dfs  카운트하면서 탐색
타겟 찾았을 때 최소 answer 가 반환 되도록 함
*/

string targ;
vector<string> word;
int size1;
int wordsize;
int answer;

void bfs(int n, string begin, vector<int> visited)
{
    if (begin.compare(targ) == 0 && n < answer) // compare 같으면 0 리턴,  다르면 -1 리턴
        answer = n;
    for (int i = 0; i < size1; i++)
    {
        if (visited[i] == 1)
            continue; // 방문 했으면 continue
        int t = 0;
        for (int j = 0; j < wordsize; j++)
        {
            if (begin[j] != word[i][j])
                t++; // 두 글자 이상 틀리면 break
            if (t > 1)
                break;
        }
        if (t == 1)
        {
            visited[i] = 1; // 한 글자 다르면 방문처리하고 dfs 진입,  카운트함, 다음 노드, 방문 넘김
            bfs(n + 1, word[i], visited);
        }
    }
}

int solution(string begin, string target, vector<string> words)
{
    answer = 999999999;
    size1 = words.size();
    wordsize = begin.length();
    targ = target;
    word = words;
    vector<int> visited(size1, 0);
    bfs(0, begin, visited);
    if (answer == 999999999)
        return 0;
    return answer;
}

int main(void)
{
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<int> dis_word_len(3, 10);  // 각 단어의 길이는 3 이상 10 이하이며 모든 단어의 길이는 같음
    uniform_int_distribution<int> dis_words_num(3, 50); // words에는 3개 이상 50개 이하의 단어가 있으면 중복되는 단어는 없음
    uniform_int_distribution<int> dis_alpha(97, 122);   // 알바벳 소문자

    int word_len = dis_word_len(gen);   // 단어의 길이
    int words_num = dis_words_num(gen); // words안에 단어의 갯수
    // words_num = 50;

    // begin 생성
    string begin = "";
    for (int i = 0; i < word_len; i++)
        begin += char(dis_alpha(gen));
    cout << "begin: " << begin << endl;

    // target 생성 (단, begin과 같지 않음)
    string target = "";
    while (true)
    {
        for (int i = 0; i < word_len; i++)
            target += char(dis_alpha(gen));

        if (target != begin)
            break;
    }
    cout << "target: " << target << endl;

    // words 생성 (중복 없음)
    vector<string> words;
    while (true)
    {
        string word = "";
        for (int j = 0; j < word_len; j++)
        {
            word += char(dis_alpha(gen));
        }
        if (find(words.begin(), words.end(), word) == words.end()) // 중복되는 단어 아님
            words.push_back(word);
        if (words.size() == words_num)
            break;
    }
    cout << "words_num: " << words_num << endl;
    cout << "words: " << endl;
    for (int i = 0; i < words_num; i++)
    {
        cout << words[i] << ' ';
    }
    cout << endl;

    begin = "hit";
    target = "cog";
    words = {"hot", "dot", "dog", "lot", "log", "cog"};

    auto start_time = chrono::steady_clock::now();
    int result = 0;
    result = solution(begin, target, words);
    auto end_time = chrono::steady_clock::now();
    chrono::duration<double> elapsed_seconds = end_time - start_time;
    cout << "--- " << elapsed_seconds.count() << " seconds ---\n";
    cout << "result: " << result << endl;

    return 0;
}